assets/scripts/app/StatusMessage.jsx

```diff
@@ -0,0 +1,119 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { connect } from 'react-redux'
+
+import { hideStatusMessage } from './status_message'
+import { registerKeypress, deregisterKeypress } from './keypress'
+import { URL_SIGN_IN_REDIRECT } from './routing'
+import { undo } from '../streets/undo_stack'
+import { loseAnyFocus } from './focus'
+import { t } from '../app/locale'
+
+const STATUS_MESSAGE_HIDE_DELAY = 15000
+
+// TODO: Some logic needs to separate out as container vs presentational.
+class StatusMessage extends React.PureComponent {
+  constructor (props) {
+    super(props)
+
+    this.timerId = -1
+  }
+
+  componentDidMount () {
+    // As per issue #306.
+    window.addEventListener('stmx:save_street', hideStatusMessage)
+  }
+
+  componentDidUpdate (prevProps) {
+    // Any update resets the visibility timer
+    window.clearTimeout(this.timerId)
+
+    // Whenever visibility is set to true, start the timer to auto-hide this
+    if (this.props.visible === true) {
+      this.timerId = window.setTimeout(hideStatusMessage, STATUS_MESSAGE_HIDE_DELAY)
+    }
+
+    // Whenever the visibility state flips, set up or tear down a keypress
+    // listener to show/hide the status message.
+    if (prevProps.visible === false && this.props.visible === true) {
+      registerKeypress('esc', hideStatusMessage)
+    } else if (prevProps.visible === true && this.props.visible === false) {
+      deregisterKeypress('esc', hideStatusMessage)
+    }
+  }
+
+  onClickUndo (event) {
+    undo()
+  }
+
+  onClickTheX (event) {
+    hideStatusMessage()
```

Comment:
I think we might have talked about this in the past but I can't remember. What do you think about using mapDispatchToProps instead of having the actions be dispatched in another file?